src/meta/backup_engine.cpp

```diff
@@ -74,9 +75,128 @@ error_code backup_engine::set_block_service(const std::string &provider)
+                          err = resp.err;
+                          remote_file = resp.file_handle;
+                      })
+        ->wait();
+    if (err != dsn::ERR_OK) {
+        ddebug_f("create file {} failed", file_name);
+        return err;
+    }
+    dassert_f(remote_file != nullptr,
+              "create file {} succeed, but can't get handle",
+              create_file_req.file_name);
+    remote_file
+        ->write(dist::block_service::write_request{write_buffer},
+                TASK_CODE_EXEC_INLINED,
+                [&err](const dist::block_service::write_response &resp) { err = resp.err; })
+        ->wait();
+    return err;
+}
+
+error_code backup_engine::backup_app_meta()
+{
+    std::shared_ptr<app_state> app;
+    error_code err = get_app_state(_cur_backup.app_id, app);
+    if (err != ERR_OK) {
+        return err;
+    }
+    // do not write app envs to backup file
+    app_state tmp = *app;
+    tmp.envs.clear();
+    dsn::blob buffer = dsn::json::json_forwarder<app_info>::encode(tmp);
+    std::string file_name = cold_backup::get_app_metadata_file(_backup_service->backup_root(),
+                                                               _cur_backup.app_name,
+                                                               _cur_backup.app_id,
+                                                               _cur_backup.backup_id);
+    return write_backup_file(file_name, buffer);
+}
+
+void backup_engine::backup_app_partition(const gpid &pid)
+{
+    std::shared_ptr<app_state> app;
+    error_code err = get_app_state(pid.get_app_id(), app);
+    if (err != ERR_OK) {
+        zauto_lock lock(_lock);
+        is_backup_failed = true;
+        return;
+    }
+
+    dsn::rpc_address partition_primary = app->partitions[pid.get_partition_index()].primary;
+    if (partition_primary.is_invalid()) {
+        dwarn_f("backup_id({}): partition {} doesn't have a primary now, retry to backup it later.",
+                _cur_backup.backup_id,
+                pid.to_string());
+        tasking::enqueue(LPC_DEFAULT_CALLBACK,
+                         &_tracker,
+                         [this, pid]() { backup_app_partition(pid); },
+                         0,
+                         std::chrono::seconds(10));
+        return;
+    }
+
+    auto req = std::make_unique<backup_request>();
+    req->pid = pid;
+    policy_info backup_policy_info;
+    backup_policy_info.__set_backup_provider_type(_provider_type);
+    backup_policy_info.__set_policy_name(get_policy_name());
+    req->policy = backup_policy_info;
+    req->backup_id = _cur_backup.backup_id;
+    req->app_name = _cur_backup.app_name;
+
+    ddebug_f("backup_id({}): send backup request to partition {}, target_addr = {}",
+             _cur_backup.backup_id,
+             pid.to_string(),
+             partition_primary.to_string());
+    backup_rpc rpc(std::move(req), RPC_COLD_BACKUP, 0_ms, 0, pid.thread_hash());
```

Comment:
You still need to set a timeout for this RPC and do retry-sort-of-thing if the RPC is unresponsive. Consider giving it 10s.

Reply:
Already set a 10s timeout, `on_backup_reply` would handle rpc errors.